pkg/providers/cloudstack/apibuilder.go

```diff
@@ -0,0 +1,113 @@
+	return clusterapi.MachineDeployment(clusterSpec, workerNodeGroupConfig, kubeadmConfigTemplate, cloudstackMachineTemplate)
+}
+
+// MachineDeployments returns generated CAPI MachineDeployment objects for a given cluster spec.
+func MachineDeployments(clusterSpec *cluster.Spec, kubeadmConfigTemplates map[string]*bootstrapv1.KubeadmConfigTemplate, machineTemplates map[string]*cloudstackv1.CloudStackMachineTemplate) map[string]*clusterv1.MachineDeployment {
+	m := make(map[string]*clusterv1.MachineDeployment, len(clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
+
+	for _, workerNodeGroupConfig := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		deployment := machineDeployment(clusterSpec, workerNodeGroupConfig,
+			kubeadmConfigTemplates[workerNodeGroupConfig.Name],
+			machineTemplates[workerNodeGroupConfig.Name],
+		)
+		m[workerNodeGroupConfig.Name] = &deployment
+	}
+	return m
+}
+
+func fillMachineTemplateAnnotations(machineConfig *v1alpha1.CloudStackMachineConfigSpec) map[string]string {
+	annotations := make(map[string]string, 0)
+	if machineConfig.DiskOffering != nil {
+		annotations[fmt.Sprintf("mountpath.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.MountPath
+		annotations[fmt.Sprintf("device.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Device
+		annotations[fmt.Sprintf("filesystem.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Filesystem
+		annotations[fmt.Sprintf("label.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Label
+	}
+
+	if machineConfig.Symlinks != nil {
+		links := make([]string, 0)
+		for key := range machineConfig.Symlinks {
+			links = append(links, fmt.Sprintf("%s:%s", key, machineConfig.Symlinks[key]))
+		}
+		// sorting for unit test determinism
+		sort.Strings(links)
+		annotations[fmt.Sprintf("symlinks.%s", constants.CloudstackAnnotationSuffix)] = strings.Join(links, ",")
+	}
+
+	return annotations
+}
+
+func setDiskOffering(machineConfig *v1alpha1.CloudStackMachineConfigSpec, template *cloudstackv1.CloudStackMachineTemplate) {
+	if machineConfig.DiskOffering == nil {
+		return
+	}
+
+	template.Spec.Spec.Spec.DiskOffering = cloudstackv1.CloudStackResourceDiskOffering{
+		CloudStackResourceIdentifier: cloudstackv1.CloudStackResourceIdentifier{
+			ID:   machineConfig.DiskOffering.Id,
+			Name: machineConfig.DiskOffering.Name,
+		},
+		CustomSize: machineConfig.DiskOffering.CustomSize,
+		MountPath:  machineConfig.DiskOffering.MountPath,
+		Device:     machineConfig.DiskOffering.Device,
+		Filesystem: machineConfig.DiskOffering.Filesystem,
+		Label:      machineConfig.DiskOffering.Label,
+	}
+}
+
+// MachineTemplate returns a generated CloudStackMachineTemplate object for a given EKS-A CloudStackMachineConfig.
+func MachineTemplate(name string, machineConfig *v1alpha1.CloudStackMachineConfigSpec) *cloudstackv1.CloudStackMachineTemplate {
```

Comment:
curious why pass the `CloudStackMachineConfigSpec` instead of the `CloudStackMachineConfig`?

Reply:
The main reason was encapsulation. There's no need for this method to have access to the whole CloudStackMachineConfig - it only needs the spec

Reply:
I see
I prefer to pass the machine templates for consistency. The spec itself misses some context, it doesn't really represent anything by itself. I understand not passing big things like the cluster.Spec, but in this case it seems we lose more than we win.

That said, I don't have big concerns, I'll leave it up to you, lgtm